core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -491,6 +512,30 @@ public interface JoinLeftEmptyRuleConfig extends PruneEmptyRule.Config {
     }
   }
 
+  private static void addNullLiterals(RexBuilder rexBuilder, Values empty,
+      List<RexNode> projectFields, List<String> newColumnNames) {
+    for (int i = 0; i < empty.getRowType().getFieldList().size(); ++i) {
+      RelDataTypeField relDataTypeField = empty.getRowType().getFieldList().get(i);
+      RexNode nullLiteral = rexBuilder.makeNullLiteral(relDataTypeField.getType());
+      projectFields.add(nullLiteral);
+      newColumnNames.add(empty.getRowType().getFieldList().get(i).getName());
+    }
+  }
+
+  public static void copyProjects(RexBuilder rexBuilder, RelDataType inRowType,
+      RelDataType castRowType, int castRowTypeOffset,
+      List<RexNode> outProjects, List<String> outProjectNames) {
+
```

Comment:
Nitpick: could you please remove this empty line to keep it consistent with the rest of the code?

Reply:
removed